### PR TITLE
feat: improve error messages when signing and verifying messages

### LINKED ISSUE #2335: Fix the `forest-cli wallet sign|verify` commands
**Issue summary**

We should be able to sign and verify messages using the wallet keys. However, something seems to have broken.

Steps to reproduce:
1. Boot calibnet and sync to HEAD.
2. Get an address from your wallet: `forest-cli --token {token} wallet list`
3. Sign a message: `forest-cli --token {token} wallet sign -a {address} -m {message}`
4. Verify a signature: `forest-cli --token {token} wallet verify -a {address} -m {message} -s {signature}`

Current, incorrect output:
```bash
$ ./target/release/forest-cli --token {token} wallet sign -a f1uatkwrkzfn5yuw67sxoosh37rujleu6xvb2dhwq -m "test message"
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidHexCharacter { c: 't', index: 0 }', forest/cli/src/cli/wallet_cmd.rs:239:52
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

```bash
$ ./target/release/forest-cli --token {token} wallet verify -a f1uatkwrkzfn5yuw67sxoosh37rujleu6xvb2dhwq -m "test message" -s "dummy signature"
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OddLength', forest/cli/src/cli/wallet_cmd.rs:256:56
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
**Summary of changes**
Changes introduced in this pull request:
- 

**Reference issue to close (if applicable)**

Closes #2335

**Other information and links**

After
```
Error: Message has to be a hex string

Caused by:
    Invalid character 't' at position 0
```

Before
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidHexCharacter { c: 't', index: 0 }', forest/cli/src/cli/wallet_cmd.rs:239:52
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### CODE REVIEW

**Comment by lemmih** [comment]
Show before and after examples of the error message. And keep in mind that panics may not be used for user-facing error messages.

File: forest/cli/src/cli/wallet_cmd.rs [hunk 1]
```diff
@@ -236,7 +237,9 @@ impl WalletCommands {
 
                 let address = address_result.unwrap();
 
-                let message = hex::decode(message).unwrap();
+                let message = hex::decode(message)
+                    .context("Message has to be a hex string")
+                    .unwrap();
```
Comment (lemmih) [inline forest/cli/src/cli/wallet_cmd.rs:242]: `unwrap` isn't a good way of communicating error messages to users.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/cli/src/cli/db_cmd.rs [hunk 2]
```diff
@@ -20,15 +20,15 @@ pub enum DBCommands {
...
+    pub async fn run(&self, config: &Config) -> anyhow::Result<()> {
         match self {
             Self::Stats => {
                 use human_repr::HumanCount;
 
                 let dir = db_path(config);
                 println!("Database path: {}", dir.display());
                 let size = fs_extra::dir::get_size(dir).unwrap_or_default();
-                println!("Database size: {}", size.human_count_bytes());
+                Ok(println!("Database size: {}", size.human_count_bytes()))
```
Comment (lemmih) [inline forest/cli/src/cli/db_cmd.rs:31]: It's a coincidence that `println!` happens to return the same type as us. Let's separate these lines to make it clear that we're not actually returning something from `println!`.

**Review by lemmih — CHANGES_REQUESTED** [review]
- Put the before and after texts in the PR description.
- Change `handle_rpc_err` to map between `JsonRpcError` and `anyhow::Error`. That way, it doesn't change the control-flow unexpectedly.

File: forest/cli/src/cli/mod.rs [hunk 3]
```diff
@@ -121,10 +122,10 @@ pub(super) fn handle_rpc_err(e: JsonRpcError) -> ! {
             error!("JSON RPC Error: Code: {code} Message: {message}")
         }
     };
```
Comment (lemmih) [inline forest/cli/src/cli/mod.rs:124]: This function shouldn't do any logging.

File: forest/cli/src/cli/auth_cmd.rs [hunk 4]
```diff
@@ -37,36 +37,35 @@ fn process_perms(perm: String) -> Result<Vec<String>, JsonRpcError> {
...
+                    .map_err(handle_rpc_err)?;
+                let mut addr = Multiaddr::empty();
+                addr.push(config.client.rpc_address.ip().into());
+                addr.push(Protocol::Tcp(config.client.rpc_address.port()));
+                addr.push(Protocol::Http);
+                println!(
+                    "FULLNODE_API_INFO=\"{}:{}\"",
+                    String::from_utf8(token)
+                        .map_err(|e| handle_rpc_err(e.into()))
+                        .unwrap(),
```
Comment (lemmih) [inline forest/cli/src/cli/auth_cmd.rs:65]:
```suggestion
                        ?,
```

File: forest/cli/src/cli/mod.rs [hunk 5]
```diff
@@ -140,51 +141,46 @@ pub(super) fn cli_error_and_die(msg: impl AsRef<str>, code: i32) -> ! {
...
-pub(super) fn print_rpc_res_pretty<T: Serialize>(res: Result<T, JsonRpcError>) {
-    match res {
-        Ok(obj) => println!("{}", serde_json::to_string_pretty(&obj).unwrap()),
-        Err(err) => handle_rpc_err(err),
-    };
+pub(super) fn print_rpc_res_pretty<T: Serialize>(
+    res: Result<T, JsonRpcError>,
+) -> anyhow::Result<()> {
+    let obj = res.map_err(handle_rpc_err)?;
+    println!("{}", serde_json::to_string_pretty(&obj).unwrap());
```
Comment (lemmih) [inline forest/cli/src/cli/mod.rs:155]:
```suggestion
    println!("{}", serde_json::to_string_pretty(&obj)?);
```

File: forest/cli/src/cli/snapshot_cmd.rs [hunk 6]
```diff
@@ -175,6 +175,7 @@ impl SnapshotCommands {
                     .unwrap();
```
Comment (lemmih) [inline forest/cli/src/cli/snapshot_cmd.rs:175]:
```suggestion
                    ?;
```

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/cli/src/cli/mod.rs (hunk 3 above)
Comment (tyshko5) [reply forest/cli/src/cli/mod.rs:124]: Removed

File: forest/cli/src/cli/auth_cmd.rs (hunk 4 above)
Comment (tyshko5) [reply forest/cli/src/cli/auth_cmd.rs:65]: Fixed

File: forest/cli/src/cli/mod.rs (hunk 5 above)
Comment (tyshko5) [reply forest/cli/src/cli/mod.rs:155]: Fixed

File: forest/cli/src/cli/snapshot_cmd.rs (hunk 6 above)
Comment (tyshko5) [reply forest/cli/src/cli/snapshot_cmd.rs:175]: Fixed

**Review by lemmih — CHANGES_REQUESTED** [review]
This is the hex error when signing:
```
Error: Message has to be a hex string

Caused by:
    Invalid character 'l' at position 1
```

This is the hex error when verifying:
```
Error: {"code":0,"message":"Invalid character 'l' at position 1"}
```

Let's make things more consistent by moving the hex-decoding from the node (aka `wallet_api.rs`) to the client (aka `wallet_cmd.rs`).

File: forest/cli/src/cli/wallet_cmd.rs [hunk 7]
```diff
@@ -237,7 +237,7 @@ impl WalletCommands {
 
                 let address = address_result?;
 
-                let message = hex::decode(message).context("Message has to be a hex string")?;
+                let message = hex::decode(message)?;
```
Comment (lemmih) [inline forest/cli/src/cli/wallet_cmd.rs:240]: Why did you remove this context?

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/cli/src/cli/wallet_cmd.rs (hunk 7 above)
Comment (tyshko5) [reply forest/cli/src/cli/wallet_cmd.rs]: Accident. Reverted

File: node/rpc/src/wallet_api.rs [hunk 8]
```diff
@@ -271,9 +271,8 @@ where
     DB: Blockstore,
     B: Beacon,
 {
-    let (addr_str, msg_str, SignatureJson(sig)) = params;
+    let (addr_str, msg, SignatureJson(sig)) = params;
```
Comment (lemmih) [inline node/rpc/src/wallet_api.rs:274]: Let's use `AddressJson` like in the sign endpoint.

File: forest/cli/src/cli/wallet_cmd.rs [hunk 9]
```diff
@@ -234,46 +235,44 @@ impl WalletCommands {
...
-                .unwrap();
+                .map_err(handle_rpc_err)?;
                 println!("{}", hex::encode(response.0.bytes()));
+                Ok(())
             }
             Self::Verify {
                 message,
                 address,
                 signature,
             } => {
```
Comment (lemmih) [inline forest/cli/src/cli/wallet_cmd.rs:256]: Let's parse the `Address`  and use `address.protocol()` to implement this logic more robustly.

**Review by lemmih — CHANGES_REQUESTED** [review]

**Comment by lemmih** [comment]
The overall goal is to have identical error messages in `sign` and `verify`. Right now the error messages are the same for hex messages but not for invalid addresses.

File: forest/cli/src/cli/wallet_cmd.rs (hunk 9 above)
Comment (tyshko5) [reply forest/cli/src/cli/wallet_cmd.rs:256]: First is done. By second do you mean usage of `state_account_key`?

Comment (lemmih) [reply forest/cli/src/cli/wallet_cmd.rs:256]: No, look at how the signature is defined. It can be done much better if you use [protocol](https://docs.rs/fvm_shared/2.0.0/fvm_shared/address/struct.Address.html#method.protocol).

Comment (tyshko5) [reply forest/cli/src/cli/wallet_cmd.rs:256]: Pushed. Hope that's what expected

**Review by lemmih — APPROVED** [review]

**Review by hanabi1224 — APPROVED** [review]
